vpr/src/route/edge_groups.h

```diff
@@ -0,0 +1,50 @@
+#ifndef EDGE_GROUPS_H
+#define EDGE_GROUPS_H
+
+#include <unordered_set>
+#include <unordered_map>
+#include <vector>
+#include <cstddef>
+
+#include "vpr_types.h"
+#include "vpr_context.h"
+
+// Class for building a group of connected edges.
+class EdgeGroups {
+  public:
+    EdgeGroups() {}
+
+    // Adds non-configurable (undirected) edge to be grouped.
+    //
+    // Returns true if this is a new edge.
+    bool add_non_config_edge(int from_node, int to_node);
+
+    // After add_non_config_edge has been called for all edges, create_sets
+    // will form groups of nodes that are connected via non-configurable
+    // edges.
+    void create_sets();
+
+    // Create t_non_configurable_rr_sets from set data.
+    // NOTE: The stored graph is undirected, so this may generate reverse edges that don't exist.
+    t_non_configurable_rr_sets output_sets();
+
+    // Set device context structures for non-configurable node sets.
+    void set_device_context(DeviceContext& device_ctx);
+
+  private:
```

Comment:
Need to comment how these data structures work / are used.  
I believe the graph stores entries for all rr_nodes that have at least one non-configurable edge. Each rr_node involved as either a source or destination of a non-configurable edge is put into the graph, with the edge storing the rr_node id of the other end of the edge.  Access the graph by rr_node id, and you will get back a set of the non-configurable edges (ids of the rr_node on the other side of the edge).  I'm still not sure what the int set value is encoding (what set number is this rr_node part of)?

Reply:
Done.